libopenage/screenshot.cpp

```diff
@@ -46,44 +52,87 @@ std::string ScreenshotManager::gen_next_filename() {
 
 
 void ScreenshotManager::save_screenshot() {
+	coord::pixel_t width = this->window_size.x,
+	               height = this->window_size.y;
+
+	auto pxdata = std::make_unique<uint32_t[]>(width * height);
+	glReadPixels(0, 0, width, height, GL_RGBA, GL_UNSIGNED_BYTE, pxdata.get());
+
+	encode_png(std::move(pxdata), width, height);
+
+	auto enc_no_args = std::bind(encode_png, std::move(pxdata), width, height);
```

Comment:
This is invalid. the `pxdata` ownage was transferred, the object is now invalid. I think you just have to remove the above function and replace the `std::bind` with:

```
auto encode_func = [pixels=std::move(pxdata), width, height] () mutable {
    encode_png(std::move(pixels), width, height);
};
```

This method should be documented in the jobmanager howto ;)

Reply:
still doesn't work. does encode_png need to be static?:
```
/home/mcd/.repo/ap/openage/libopenage/screenshot.cpp: In lambda function:
/home/mcd/.repo/ap/openage/libopenage/screenshot.cpp:64:53: error: ‘this’ was not captured for this lambda function
   return encode_png(std::move(pixels), width, height);
                                                     ^
/home/mcd/.repo/ap/openage/libopenage/screenshot.cpp:64:53: error: cannot call member function ‘bool openage::ScreenshotManager::encode_png(std::unique_ptr<unsigned char []>&&, openage::coord::pixel_t, openage::coord::pixel_t)’ without object
/home/mcd/.repo/ap/openage/libopenage/screenshot.cpp: In member function ‘void openage::ScreenshotManager::save_screenshot()’:
/home/mcd/.repo/ap/openage/libopenage/screenshot.cpp:66:50: error: no matching function for call to ‘openage::job::JobManager::enqueue(openage::ScreenshotManager::save_screenshot()::<lambda()>&)’
  this->job_manager->enqueue<bool>(encode_function);
                                                  ^
In file included from /home/mcd/.repo/ap/openage/libopenage/screenshot.h:11:0,
                 from /home/mcd/.repo/ap/openage/libopenage/screenshot.cpp:3:
/home/mcd/.repo/ap/openage/libopenage/job/job_manager.h:93:9: note: candidate: template<class T> openage::job::Job<T> openage::job::JobManager::enqueue(openage::job::job_function_t<T>, openage::job::callback_function_t<T>)
  Job<T> enqueue(job_function_t<T> function,
         ^
/home/mcd/.repo/ap/openage/libopenage/job/job_manager.h:93:9: note:   template argument deduction/substitution failed:
/home/mcd/.repo/ap/openage/libopenage/screenshot.cpp:66:50: note:   cannot convert ‘encode_function’ (type ‘openage::ScreenshotManager::save_screenshot()::<lambda()>’) to type ‘openage::job::job_function_t<bool> {aka std::function<bool()>}’
  this->job_manager->enqueue<bool>(encode_function);
                                                  ^
In file included from /home/mcd/.repo/ap/openage/libopenage/screenshot.h:11:0,
                 from /home/mcd/.repo/ap/openage/libopenage/screenshot.cpp:3:
/home/mcd/.repo/ap/openage/libopenage/job/job_manager.h:113:9: note: candidate: template<class T> openage::job::Job<T> openage::job::JobManager::enqueue(openage::job::abortable_function_t<T>, openage::job::callback_function_t<T>)
  Job<T> enqueue(abortable_function_t<T> function,
         ^
/home/mcd/.repo/ap/openage/libopenage/job/job_manager.h:113:9: note:   template argument deduction/substitution failed:
/home/mcd/.repo/ap/openage/libopenage/screenshot.cpp:66:50: note:   cannot convert ‘encode_function’ (type ‘openage::ScreenshotManager::save_screenshot()::<lambda()>’) to type ‘openage::job::abortable_function_t<bool> {aka std::function<bool(std::function<bool()>, std::function<void()>)>}’
  this->job_manager->enqueue<bool>(encode_function);
                                                  ^

```

Reply:
add `this` inside the `[]`, so you have `... = [this, pixels=std::move(pxdata), width, height] ()...`, and call `this->encode_png`. Or make `encode_png` static, if it doesn't need access to `this`.

Reply:
we're getting closer. It's nice gcc tells me that the problem is, that the function is implicitly deleted because else the default definition would be ill-formed. but why would it be ill-formed, gcc?:
```
/home/mcd/.repo/ap/openage/libopenage/screenshot.cpp: In member function ‘void openage::ScreenshotManager::save_screenshot()’:
/home/mcd/.repo/ap/openage/libopenage/screenshot.cpp:64:50: error: use of deleted function ‘openage::ScreenshotManager::save_screenshot()::<lambda()>::<lambda>(const openage::ScreenshotManager::save_screenshot()::<lambda()>&)’
  this->job_manager->enqueue<bool>(encode_function);
                                                  ^
/home/mcd/.repo/ap/openage/libopenage/screenshot.cpp:61:71: note: ‘openage::ScreenshotManager::save_screenshot()::<lambda()>::<lambda>(const openage::ScreenshotManager::save_screenshot()::<lambda()>&)’ is implicitly deleted because the default definition would be ill-formed:
  auto encode_function = [this, pixels=std::move(pxdata), width, height] () mutable {
                                                                       ^
/home/mcd/.repo/ap/openage/libopenage/screenshot.cpp:61:71: error: use of deleted function ‘std::unique_ptr<_Tp [], _Dp>::unique_ptr(const std::unique_ptr<_Tp [], _Dp>&) [with _Tp = unsigned char; _Dp = std::default_delete<unsigned char []>]’
In file included from /usr/include/c++/5/memory:81:0,
                 from /home/mcd/.repo/ap/openage/libopenage/screenshot.h:7,
                 from /home/mcd/.repo/ap/openage/libopenage/screenshot.cpp:3:
/usr/include/c++/5/bits/unique_ptr.h:598:7: note: declared here
       unique_ptr(const unique_ptr&) = delete;
       ^
In file included from /usr/include/c++/5/memory:79:0,
                 from /home/mcd/.repo/ap/openage/libopenage/screenshot.h:7,
                 from /home/mcd/.repo/ap/openage/libopenage/screenshot.cpp:3:
/usr/include/c++/5/functional:2246:7: note:   initializing argument 1 of ‘std::function<_Res(_ArgTypes ...)>::function(_Functor) [with _Functor = openage::ScreenshotManager::save_screenshot()::<lambda()>; <template-parameter-2-2> = void; <template-parameter-2-3> = void; _Res = bool; _ArgTypes = {}]’
       function<_Res(_ArgTypes...)>::
       ^
/usr/include/c++/5/functional: In instantiation of ‘static void std::_Function_base::_Base_manager<_Functor>::_M_clone(std::_Any_data&, const std::_Any_data&, std::false_type) [with _Functor = openage::ScreenshotManager::save_screenshot()::<lambda()>; std::false_type = std::integral_constant<bool, false>]’:
/usr/include/c++/5/functional:1746:16:   required from ‘static bool std::_Function_base::_Base_manager<_Functor>::_M_manager(std::_Any_data&, const std::_Any_data&, std::_Manager_operation) [with _Functor = openage::ScreenshotManager::save_screenshot()::<lambda()>]’
/usr/include/c++/5/functional:2256:19:   required from ‘std::function<_Res(_ArgTypes ...)>::function(_Functor) [with _Functor = openage::ScreenshotManager::save_screenshot()::<lambda()>; <template-parameter-2-2> = void; <template-parameter-2-3> = void; _Res = bool; _ArgTypes = {}]’
/home/mcd/.repo/ap/openage/libopenage/screenshot.cpp:64:50:   required from here
/usr/include/c++/5/functional:1710:34: error: use of deleted function ‘openage::ScreenshotManager::save_screenshot()::<lambda()>::<lambda>(const openage::ScreenshotManager::save_screenshot()::<lambda()>&)’
    __dest._M_access<_Functor*>() =
                                  ^
```

Reply:
I think the the explaination is that the lambda function's copy constructor is deleted because it captures a unique_ptr. If one could copy the lambda, you would also copy the contained unique_ptr. The solution is to add a `JobManager::enqueue(job_function_t<T> &&)` method, otherwise this lambda will always be copied into the jobmanager, which is wrong.

This would mean some more complication, so I'd just go for a shared_ptr. It behaves exactly the same, except you can copy it.